Payments refactor

michaljelonek opened the pull request:
Payments refactor, this is still fully in progress, so a lot may and probably will change. Change one thing and it breaks in multiple other places, hard to keep track of the changes.

I still have your comments in mind @jxltom, just didn't have time (or an idea how) to implement them yet. You might want to look at `process_payment` to see a different structure that I made with them in mind. 

### Pull Request Checklist

<!-- Please keep this section. It will make maintainer's life easier. -->

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -167,75 +234,89 @@ def gateway_authorize(payment: Payment, payment_token: str) -> Transaction:
-    if not txn.is_success:
-        raise PaymentError(error)
-    return txn
+    auth_transaction = payment.transactions.filter(
```
The ```auth_transaction``` could be None and it should be validated.

michaljelonek replied:
My understanding was that this was already provided by the functions called above this. Like 'clean_capture' calls 'payment.can_capture' which in turn has a check for 'is_authorized'.

jxltom replied:
If so, these helper needs to be updated. As far as I know, they didn't have checks on the transactions. They only check the payment status or charge amount.

Plus, there is no harm to recheck this to avoid exceptions.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -167,75 +234,89 @@ def gateway_authorize(payment: Payment, payment_token: str) -> Transaction:
-    if not txn.is_success:
-        raise PaymentError(error)
-    return txn
+    transaction = payment.transactions.filter(
```
Same, the transaction could be ```None``` and should be validated.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -167,75 +234,89 @@ def gateway_authorize(payment: Payment, payment_token: str) -> Transaction:
-        raise PaymentError(error)
-    return txn
+
+    auth_transaction = payment.transactions.filter(
```
Same, the transaction could be ```None``` and should be validated.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
+    """Validates response to be a correct format for Saleor to process."""
+    if not isinstance(response, dict):
+        raise GatewayError('Gateway needs to return a dictionary')
+    required_fields = {'transaction_id', 'is_success', 'errors'}
```
It's not required for Saleor to work, you can send it, yes, but I wouldn't go as far as to make it a required field. Also you can send back whatever you want, the whole response returned from gateway is saved, not a specific key from the dictionary, this makes it easier as you can also return additional data and not worry about proper key naming. Notice I'm not raising a validation error if there's more returned than those 3 keys. Just as long as it's encodable by DjangoJsonEncoder, which I should probably check here just in case.

jxltom replied:
If required means required for Saleor to work, I can accept it. :D

michaljelonek replied:
I might've used wrong wording, but the `required_keys` is something that, in my mind, is required in order to process payments, `raw_response` is not needed in order to do.

jxltom replied:
Just to elaborate it, raw response is not crucial to saleor but it is essential to a payment system. If someone write a provider without saving the raw response, I probably won't use it. The payment gateway is related with real money, and we should take it seriously. Without raw response, we can barely debug the transaction issues if any issues happens and this is a big problem.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -68,23 +73,18 @@ def create_payment(**payment_data):
+        is_success=gateway_response.get('is_success', False),
+        amount=extra_params.get('amount', payment.total),
+        currency=payment.currency,
         gateway_response=gateway_response)
```
See my other comment, doesn't really matter what the key is named, the whole response is saved. Whether the gateway wants to send one is up to the gateway :P

jxltom replied:
Yeah, but it will be nice to keep it as a convention. You are using ```raw_response``` in braintree but ```gateway_response``` in dummy. Why not just keep them same as ```raw_response```.

jxltom replied:
In this case it is better removing ```gateway_response``` from dummy, otherwise it is confusing. And the docs should have explanations for this gateway response.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
+            **extra_params, **gateway_params)
+        validate_gateway_response(gateway_response)
+    except AttributeError:
+        logger.exception('Gateway doesn\'t implement %s', func_name)
```
Fair.

michaljelonek commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -136,128 +118,111 @@ def get_client_token(**connection_params: Dict) -> str:
+    return {
+        'is_success': result.is_success,
+        'transaction_id': gateway_response['transaction_id'],
+        'raw_response': gateway_response,
```
I don't agree, why make a response just for the sake of having a response?

jxltom replied:
Because this is a little bit confusing. There should still be a raw response though it is a dummy gateway. You are wring a standard, if this is not kept as a convention, it requires docs explaining for that. So why not do it in code.

michaljelonek replied:
The docs will be updated to match, and everything will be explained. Certain gateways will be provided as examples as well. Dummy.. I'd say is not a good example, it's basically a debug gateway to just force the payment into a certain state. Besides currently it's kinda broken :D

jxltom replied:
That's will be great.

michaljelonek commented on saleor/order/views.py:
```diff
@@ -107,24 +107,25 @@ def start_payment(request, order, gateway):
-                form.process_payment()
-            except PaymentError as exc:
-                form.add_error(None, exc.message)
+                gateway_process_payment(payment=payment,
```
It is better, because the form is then decoupled from the actual logic of the gateway. Why make a form, that should theoretically validate and handle the data, call gateway logic? Now this would be okay if it called Saleor, so Saleor can handle the payment and call gateway methods, but that's impossible if we want to make gateways separate of Saleor. Of course the form could call the gateway methods directly, but that would then split the logic into multiple places and to gain what? It makes it harder to debug and more confusing. It's better to keep the form minimal and return only required stuff for the payment to work (in this case so far it's been the payment token).

jxltom replied:
It makes sense. I'm only afraid that some gateway may not have a way to ```get_payment_token```. But it works so far to the current three payment providers.

michaljelonek replied:
Yeah, I was worried about that as well, but so far we don't have a case like that and we might never have actually. At the point where this becomes an issue I think payments will be simpler, so implementing such changes to the form and the gateways should be easier as well.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
+    finally:
+        transaction = create_transaction(
+            payment=payment,
+            kind=transaction_kind,
```
Is this only the case with auth (in regards of capture/charge)? Cause I'd rather not make the Saleor flow dependable on the gateway. I'd rather have it the other way, so Saleor enforces some kind of flow and the gateways have to conform to it. I agree with the transaction skip though, but that can be solved with a simple additional key in the response (like 'was_authorized: True/False') that could then create an additional auth transaction.

jxltom replied:
As far as I know this is the only case.

An additional key may not be enough. If it is truly a separation operation in third-party gateway, we should record all the info includes raw response, amount, currency, etc.

michaljelonek replied:
This means that returning a list might be unavoidable.. Kinda sad if it comes to that, I like how simple it is, but this will make it more complex just to conform to that one case of two operations. I wonder if there's a better solution.

jxltom replied:
I know this looks not good. But recording all the information to a payment is more important.

Maybe we could consider supporting both a list or a dict? For normal case, ```call_gateway``` will take a dict as returned value, but if it is a list, the ```call_gateway``` should also be able to handle it.

michaljelonek replied:
I think I'd prefer at least enforcing a list on the Saleor side. Sure, the gateway can return a dict for simplicity, but on Saleor I'd have it made a list if it's a dict to ensure that we have one logic (of say iterating and creating transactions). I'll go with this, I'm a bit sceptical having the gateway return what kind of transaction it did, but sure, let's go for it :)

jxltom replied:
That's better. At least mypy could check the types easily.

Maybe it worth considering let the gateway itself returning the transaction kind and use it directly in creating transactions.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
         raise PaymentError('Charged transactions cannot be authorized again.')
 
 
+def call_gateway(
```
That's the thing and the reason why something like `process_payment` exists, Saleor has no way of knowing what the gateway implements. Otherwise we could just call `auth`, then `capture` every time and what the gateway does.. we essentially don't care. But I feel that would be worse than making `process_payment` method, that we expect to actually process the payment (be it auth and .capture or charge). I don't really dig returning a list, but we might end up doing that in the end.

jxltom replied:
Why not let the gateway itself returns a list of gateway response with KIND information. Then ```process_payment``` only need to iterate it and create different transactions.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
+    try:
+        gateway_response = getattr(gateway, func_name)(
+            payment=model_to_dict(payment), payment_token=payment_token,
+            **extra_params, **gateway_params)
```
In this case, I would assume ```amount``` in gateway APIs such as ```capture```, ```void```, etc. can be safely used directly.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -68,23 +73,18 @@ def create_payment(**payment_data):
-        currency=currency,
+        token=gateway_response.get('transaction_id', payment_token),
+        is_success=gateway_response.get('is_success', False),
+        amount=extra_params.get('amount', payment.total),
```
So you're saying even though we ask the gateway to say charge X amount, there's no certainty of that and the gateway can end up charging less/more? That's a bit inconvenient, but if that's the case then indeed the amount has to be provided in the response. Although we can't really validate that for all methods, can we? Not everything returns money, for example void doesn't work on amount, does it?

jxltom replied:
Yes, it is very likely for some gateways, and the gateway will actually return success. So we can only trust the amount returned from the gateway. Othewise we may have discrepancy.

We can't make sure they are all validated in saleor's side but we should do that if we can. Some payment provider won't return the amount for some operations probably.

So in the gateway implementation, they should return ```amount``` from the raw response if they can. If they can't, they should still provide a default amount such as ```payment.total```.

jxltom replied:
For the ```charge```, usually it is fine. It happens usually in ```refund``` or ```void```.  You ask for void X, but the payment provider will void less. It is because they don't support partial void or paitial refund.

michaljelonek commented:
@jxltom great review, I added the things we discussed:
- call_gateway has support for multiple gateway responses
- dummy doesn't return raw_response anymore
- added validation for kind, amount and currency from gateway response - the trust has been moved to the gateway response
- everything still operates on the last transaction, e.g. [auth, capture] -> capture

Do take a look, there's been so many comments I might've missed something.

Next steps:
- introduce a common format of data sent to the gateway, instead of sending kwargs I'd rather send one dict that contains information in a better way, e.g.
```
payment_information = {
    'token': ...,
    'amount': ...,
    'currency': ...,
    'billing': { ... },
    'shipping': { ... }
}
```
- add the "little" things like extra validation, small fixes and improvements, make the code more readable
- refactor the other gateways to work (razorpay, braintree, stripe) with those new changes

jxltom commented on saleor/payment/utils.py:
```diff
@@ -125,9 +147,91 @@ def clean_authorize(payment: Payment):
+        for response in gateway_response:
+            transaction = create_transaction(
+                payment=payment,
+                kind=transaction_kind,
```
The ```kind``` is no longer necessary since the gateway response has to return the ```kind``` information as a required field.

michaljelonek replied:
It is required, well.. kinda, this is in case a rare situation happens that gateway explodes with an exception and doesn't return a response, so we still create a transaction of a specific type.

jxltom replied:
Makes sense!

jxltom commented on saleor/payment/utils.py:
```diff
@@ -125,9 +147,91 @@ def clean_authorize(payment: Payment):
+    """Validates response to be a correct format for Saleor to process."""
+    if not isinstance(responses, (dict, list)):
+        raise GatewayError('Gateway needs to return a dictionary or a list')
+    if not isinstance(responses, list):
```
I would recommend adding a blank line before ```if not isinstance(responses, list):``` as the first if condition is for validating and the second if condition is for post-process. They actually have different logic and should be separated.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -68,25 +93,22 @@ def create_payment(**payment_data):
-        amount=amount,
-        currency=currency,
+        kind=gateway_response.get('kind', kind),
+        token=gateway_response.get('transaction_id', payment_token),
```
As these fields are required for the gateway response, probably there is no need to provide default values.

michaljelonek replied:
Defaults are for when the validation fails or an exception happens, and we don't get the response in a format we can process. I assumed we still might want to create a transaction object that something has happened.

jxltom replied:
Makes sense.

But I'm not sure whether these default are proper not not. Probably these defaults are not the actual ones sent to the gateway. For example, I think the amount is actually passed to gateway by ```extra_params```. Maybe we should use these values which are actually used in the gateways as defaults? Othewise we may create transactions with wrong data.

jxltom replied:
That sounds nice. We should try to use same logic.

jxltom replied:
That will be nice! We should try to keep the same logic at the same place.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -125,9 +147,91 @@ def clean_authorize(payment: Payment):
+                error_msg=error_msg,
+                gateway_response=response)
+
+    if not transaction.is_success:
```
Should we check all the transactions as actually there are multiple transactions are created in the above ```for``` loop?

jxltom commented on saleor/payment/utils.py:
```diff
@@ -125,9 +147,91 @@ def clean_authorize(payment: Payment):
+        # attempt to get errors from response, if none raise a generic one
+        raise PaymentError(transaction.error or GENERIC_TRANSACTION_ERROR)
+
+    return transaction
```
Is it better that returning the last transaction explicitly if there are multiple transactions?

jxltom commented on tests/api/test_order.py:
```diff
@@ -1095,13 +1095,6 @@ def test_clean_order_void_payment():
     assert errors[0].field == 'payment'
     assert errors[0].message == 'Only pre-authorized payments can be voided'
 
-    payment.is_active = True
```
Let's add this test back if we can later.

jxltom commented:
@michaljelonek Looks very nice! I will let you know if I find anything later.

michaljelonek commented:
@jxltom all done, coverage stayed roughly the same, CI green, all issues that were fixable without magic are fixed.. Do take a look on Stripe changes though, had to modify a bit to make it work with the changes, but I didn't test it yet.

@NyanKiyoshi you might want to look at Razorpay changes as I had to modify that as well and see whether it's still what you had in mind.

NyanKiyoshi commented:
@michaljelonek thanks, I will take a look once home :+1: (Wish me luck for my final exam! :tada:)

NyanKiyoshi commented on saleor/payment/gateways/razorpay/__init__.py:
```diff
@@ -132,26 +134,23 @@ def refund(payment: Payment, amount: Decimal, **connection_params):
-        payment=payment, kind=TransactionKind.REFUND, **response)
-    return transaction, error
+    return _generate_response(
+        payment_information=payment_information, kind='refund', **response)
```
We shouldn't be setting the transaction kinds through hard-coded strings, that may make it hard to refactor, or at least to maintain Saleor.

jxltom replied:
There is a validation in ```validate_gateway_response``` by ```if response['kind'] not in ALLOWED_GATEWAY_KINDS:```. But it is better if it is not hard-coded.

michaljelonek replied:
The problem is that eventually those gateways won't live in Saleor, so they won't have access to the `TrasnactionKind` that Saleor has, so there are two ways of moving forward:
1) make a common package that implements such things
2) just hard code the strings
I took the second approach for the time being, we can always move to the other one if necessary.

jxltom replied:
I think what @NyanKiyoshi means though these transaction types won't be avaible out of saleor, it still be nice to wrapper it as an enum or whatever in each gateways. So in each gateway, it doesn't need to access the hard-coded string such ```refund``` multiple time. It only needs to access the field of that transaction kind enum. It makes it easier to refactor.

michaljelonek replied:
Ah fair, sounds good, I can do that.

NyanKiyoshi commented on saleor/payment/gateways/razorpay/__init__.py:
```diff
@@ -88,11 +87,11 @@ def get_client_token(**_):
-        payment_token: str,
-        amount: Decimal,
-        **connection_params: Dict) -> Tuple[Transaction, str]:
+def process_payment(payment_information: Dict, **connection_params) -> Dict:
```
I would put suggest to put `process_payment(...)` under `charge(...)` for better readability. The user that will read the code, won't know yet what is `charge(...)`, which forces them to look around for it.

jxltom commented on docs/guides/payments.rst:
```diff
@@ -178,52 +164,74 @@ Example
++-------------------------+----------+------------------------------------------------------------------------------------+
+| name                    | type     | description                                                                        |
++-------------------------+----------+------------------------------------------------------------------------------------+
+| ``payment_information`` | ``dict`` | Payment information, containing the tokens, amount, currency and billing.          |
```
What's the key of ```payment_information```, ```tokens```, ```amount```, ```currency``` and ```billing```? What'is the type of them such as for ```billing```. Maybe it worth to mention it in the docs, so the new users to saleor can implement new gateway easier.

michaljelonek replied:
Added example data that is sent (with comments) and an additional table with fields explained for the gateway response.

jxltom commented on docs/guides/payments.rst:
```diff
@@ -178,52 +164,74 @@ Example
++----------------------+----------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
+| name                 | type                       | description                                                                                                                              |
++----------------------+----------------------------+------------------------------------------------------------------------------------------------------------------------------------------+
+| ``gateway_response`` | ``dict`` or ``list[dict]`` | Dictionary or list of dictionaries containing details about every transaction, with ``is_success`` set to ``True`` if no error occurred. |
```
What is the required fields of ```gateway_response``` which is going to be validated in ```validate_gateway_response```? It worths to mention it in the docs.

jxltom commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -136,128 +115,109 @@ def get_client_token(**connection_params: Dict) -> str:
+    payment_information['token'] = auth_resp['transaction_id']
+
+    try:
+        assert auth_resp['is_success']
```
Please don't use assert as it should only be used in development

jxltom replied:
```
if not auth_resp['is_success']:
   return [auth_resp, void(payment_information, **connection_params)]
return [auth_resp, capture(payment_information, **connection_params)]

jxltom commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -136,128 +115,109 @@ def get_client_token(**connection_params: Dict) -> str:
-        payment.cc_brand = credit_card_data.card_type
-        payment.cc_exp_month = credit_card_data.expiration_month
-        payment.cc_exp_year = credit_card_data.expiration_year
-        payment.save(update_fields=[
```
What if we truly would like to update the payment from the gateway response? How can we do this with the new payment?

michaljelonek replied:
With the new payment it's not possible for the gateway to update it. In the future if the need arises we can extend the functionality so the gateway can update the payment. For the time being only Saleor does any updating.

jxltom commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -136,128 +115,109 @@ def get_client_token(**connection_params: Dict) -> str:
+    return {
+        'is_success': result.is_success,
+        'kind': 'auth',
+        'amount': gateway_response.get('amount'),
```
It worth to mention that the gateway should only try to provide the ```amount``` and ```currency``` if they are truely in the response. They shouldn't provide default values such as ```payment.total```. The saleor's side will take care of it by providing the default values such as by ```gateway_response.get('amount', payment_information['amount'])```

jxltom replied:
If we decide the default values of amout and currency should be provided from the gateway side, then we should take them as required fields in ```validate_gateway_response``` and do not provide default values in saleor's ```create_transaction```

michaljelonek replied:
My thinking was that the gateway always provides values for required fields. Be it from the response of the provider api or reused values from what the gateway received if there's none from api response. I added extra validation to `validate_gateway_response` with type checking as well. I also made the response consistent in all gateways.

jxltom commented on saleor/payment/gateways/braintree/errors.py:
```diff
@@ -110,3 +110,17 @@ def get_error_type(error_code):
+
+class BraintreeException(Exception):
+
+    def __init__(self, *args, **kwargs):
```
I would recommend:
```
if args or kwargs:
    return super().__init__(*args, **kwargs)
 return super().__init__(DEFAULT_ERROR_MESSAGE)
```

jxltom commented on saleor/payment/gateways/dummy/__init__.py:
```diff
@@ -19,78 +16,81 @@ def get_client_token(**connection_params):
+def process_payment(payment_information: Dict, **connection_params) -> Dict:
+    """Process the whole payment, by calling authorize and capture."""
+    auth_resp = authorize(payment_information, **connection_params)
+    if auth_resp['error']:
```
We should use ```is_success``` instead of ```error``` since the gateway may fail actually but unable to provide the errors.

jxltom commented on saleor/payment/gateways/razorpay/__init__.py:
```diff
@@ -28,32 +26,29 @@
+    return {
+        'transaction_id': data.get('id', payment_information['token']),
+        'kind': kind,
+        'amount': data.get('amount', payment_information['amount']),
```
Do we really need to provide a fallback ```payment_information['amount']``` as ```create_transaction``` has alread provided a default value for ```amount``` and ```currency```? There is a difference between braintree, razorpay and stripe.

jxltom commented on saleor/payment/gateways/stripe/__init__.py:
```diff
@@ -174,50 +134,49 @@ def _get_client(**connection_params):
-        response.get('currency', payment.currency))
+        response.get('currency', payment_information['currency']))
 
+    amount = payment_information.get('amount')
```
Same here, as the gateway provides the default amount and currency, why don't we make the amount and currency as required fields in ```validate_gateway_response``` and then then there is no need to provide the default values in ```create_transaction```.

jxltom commented on saleor/payment/utils.py:
```diff
@@ -69,24 +102,24 @@ def create_payment(**payment_data):
+        token=gateway_response.get(
+            'transaction_id', payment_information['token']),
+        is_success=gateway_response.get('is_success', False),
+        amount=gateway_response.get('amount', payment_information['amount']),
```
As amount and currency are required fields, I think there is no needs to provide the default values?

michaljelonek replied:
Added a comment before this whole block 👍

jxltom commented:
@michaljelonek Nice work. The stripe part looks great to me. Do you mind clicking resolve button for the comments as I think most of them are already resolved. This page is somwhat in a mess. :D

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +158,102 @@ def clean_authorize(payment: Payment):
+    return transactions[-1]
+
+
+def validate_gateway_response(responses):
```
Do we though? Technically that does not make the response invalid, and I think if the gateway made a successful payment in a different currency (because of reasons) we still want to keep it as a successful one instead of raising an error that marks it as unsuccessful. Perhaps log a warning if the returned currency is indeed different. What do you think?

jxltom replied:
Actually that is quite important, for example in https://github.com/mirumee/saleor/blob/3eca2f4de9bef55a8ec3ea3fcda22024995bb0a6/saleor/payment/utils.py#L190, the payment's captured amount is updated from the transaction's amount directly without checking the currency. Saleor currently only support default currency, to be honest. And then the captured amount is used to determine if a order is fully paid. So correct amount with same currency is important.

If a gateway's returned currency is not same as saleor's, maybe there will be a problem. For example, saleor may think it charges 10USD, but the gateway may charges 10EURO. But saleor calculate captured amount without considering the currency. 

But in this case, the gateway made a successfull payment indeed. Maybe it is not proper to mark it as unsuccessfully since the money is actually charged though the currency is incorrect. I agree with you, maybe let's just add warnings for now.

michaljelonek replied:
Added a warning 👍

patrys commented:
I think the biggest chunk that is missing is docs. I would love to read (and discuss) a high-level textual overview before reviewing code changes.

jxltom commented on saleor/payment/gateways/razorpay/__init__.py:
```diff
@@ -8,8 +8,6 @@
 
 from . import errors
 from ... import TransactionKind
```
This can be removed

jxltom commented on saleor/payment/utils.py:
```diff
@@ -69,26 +102,27 @@ def create_payment(**payment_data):
         gateway_response=gateway_response)
     return txn
 
-
```
This blank line shouldn't be removed.

michaljelonek commented on saleor/payment/gateways/braintree/errors.py:
```diff
@@ -122,5 +122,4 @@ class BraintreeException(Exception):
             super().__init__(*args, **kwargs)
-        else:
-            super().__init__(DEFAULT_ERROR_MESSAGE)
+        super().__init__(DEFAULT_ERROR_MESSAGE)
```
Codeclimate is raising an issue that there's explicit return in the __init__ now. Should we revert to if..else or ignore it?

jxltom replied:
Hmm, ususally ```__init__``` should return ```None```, so explicit return is not recommended. In this case it should be fine. But maybe let's revert to if-else to remove this warning since there is no warning with ```if-else```. :p

patrys commented:
@michaljelonek I feel that our current docs (and the updated ones) explain what each function does. This information fits better as docstrings in code as those can be updated when the implementation changes. I'd like to see a higher-level explanation of how everything fits together that only mentions the function names and maybe contains a flowchart or two.

Here's an example: https://help.shopify.com/en/api/guides/sell-through-the-checkout-api

If anyone needs to include graphs in docs, I recommend using [Camunda](https://camunda.com/bpmn/tool/) and including both the source BPMN file and the exported graphic.

michaljelonek commented:
@patrys sure, I can add it, but can we do so in a separate PR at a later date? This PR has been open for so long already with over 100 comments now, I'd like to move forward. While I agree that more documentation is better, the current one that we ahve should be sufficient to understand what is actually happening and why (especially from a dev standpoint). I'd propose to come back to it once everything is in place.

maarcingebala commented on saleor/payment/gateways/braintree/forms.py:
```diff
@@ -30,23 +30,12 @@ def clean(self):
         # partial-payments at this moment, error is returned instead.
         amount = cleaned_data.get('amount')
-        if amount and amount != self.payment.total:
+        if amount and amount != self.payment_information['amount']:
```
Not sure of what type is `amount` here, but if it's a numeric type (which I suppose it is) and it is equal to zero, we might accidentally enter/not enter this logic, as `bool(0)` evaluates to False. Did you check that case?

michaljelonek replied:
We technically have checks for that in a different place. There shouldn't be an attempt to charge 0 if it was changed via the form. The only case this is possible if the order is for 0, but that would be an okay case then.

maarcingebala commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -33,23 +31,31 @@
 
 
-def get_customer_data(payment: Payment) -> Dict:
+class TransactionKind:
```
Couldn't we have one `TransactionKind` definition? I see that we have multiple ones and they are identical.

michaljelonek replied:
See https://github.com/mirumee/saleor/pull/3519#discussion_r247162323 and https://github.com/mirumee/saleor/pull/3519#discussion_r247901262 (re common types).

michaljelonek commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -136,128 +123,118 @@ def get_client_token(**connection_params: Dict) -> str:
     except braintree_sdk.exceptions.NotFoundError:
-        return transaction_and_incorrect_token_error(
-            payment, kind=TransactionKind.AUTH, token=payment_token)
+        raise BraintreeException
```
Docs specify that it can be also an exception class. 

Essentially it doesn't matter whether it's custom or not. Can leave it as it, or remove that block and not catch anything and just return a generic response.. Catching was added by the original author, so I thought to keep it.

michaljelonek commented on docs/guides/payments.rst:
```diff
@@ -178,52 +164,153 @@ Example
++-------------------------+----------+------------------------------------------------------------------------------------+
+| name                    | type     | description                                                                        |
++-------------------------+----------+------------------------------------------------------------------------------------+
+| ``payment_information`` | ``dict`` | Payment information, containing the token, amount, currency and billing.           |
```
That package says that it's for 3.6 and incompatible for 3.7. We support both and additionally 3.5. Besides I asked about common types between payments (thinking ahead this won't be in Saleor after next PR..), so a custom type needs to be defined in a common package. I did not get a response, so I assumed we don't want that.

patrys replied:
I think even explicitly defined classes would work here. Here's an example from `prices`:

https://github.com/mirumee/prices/blob/cd74e1fc5c7a7b993f57736fbbdebc0adaeec189/prices/money.py#L12-L24

michaljelonek commented on saleor/order/forms.py:
```diff
@@ -44,7 +45,7 @@ def clean(self):
     def save(self):
         payment = self.cleaned_data['payment']
-        payment.void()
+        gateway_void(payment)
```
It can't. `gateway_void` calls `call_gateway` which catches everything.

michaljelonek commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -33,23 +31,31 @@
+class TransactionKind:
+    AUTH = 'auth'
+    CAPTURE = 'capture'
+    CHARGE = 'charge'
```
But gateways can perform different operations. Some do not implement authorisation at all and instead do charge, some do not implement charge and require auth and capture. If you merge it, how will you tell what actually happened? And in case if one from multiple operations failed, which one?

patrys replied:
Does it matter though? Saleor says "auth and capture" and the backend does that, even if it takes several steps. Saleor says "auth but don't capture if possible" and the backend either does just auth or proceeds to capture anyway, there's no way around this. We can't really display "actually the store is configured to use a backend that does not implement preauth" to the user during checkout.

michaljelonek replied:
Not in my eyes, no. There was a similar discussion here though https://github.com/mirumee/saleor/pull/3519#discussion_r245551018. The outcome is that if the gateway does auth and capture separately, then we want to know that. If it does only charge, then we want to know that as well. Although there's no "auth but do not capture" implemented. I still think this process_payment is a bad idea and I'd rather enforce `auth` and `capture` separately (if the gateway does not implement auth because it has charge only then who cares really? it can just do charge in capture step), but that was the outcome of the discussion.

patrys replied:
It's an implementation detail. Sure, we should generate a number of transaction items but there's no reason for either Saleor or the staff to care about implementation details of the payment process. A single store may have more than one payment method enabled and they can offer different sets of features. We can't say "oh, you can't pay after all" once the customer has entered their card details. Charging too early is an inconvenience (as in case of a cancelled order refund charges will apply) but it's still better than rejecting sales.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +160,120 @@ def clean_authorize(payment: Payment):
+    )
+
+    try:
+        gateway_response = getattr(gateway, func_name)(
```
Meaning? I should split this into two calls and two try..except blocks?

patrys replied:
I mean that (assuming this function is useful), it should accept an enum token for the operation type and it should use an `if`/`elif` switch to call the correct function.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -69,22 +102,24 @@ def create_payment(**payment_data):
+
+    # Default values for token, amount, currency are only used in cases where
+    # response from gateway was invalid or an exception occured
     txn, _ = Transaction.objects.get_or_create(
```
What instead of `_`? Double underscore? Actually I don't think an update ever happens here, but I'll double check that.

michaljelonek commented on saleor/payment/utils.py:
```diff
@@ -125,9 +160,120 @@ def clean_authorize(payment: Payment):
+    except AttributeError:
+        error_msg = 'Gateway doesn\'t implement {}'.format(func_name)
+        logger.exception(error_msg)
+    except GatewayError:
```
GatewayError is when validation of the response fails, while the generic for everything else that could possible fail (including exception raised from provider sdk that are not caught in the gateway).

michaljelonek commented on saleor/payment/gateways/braintree/__init__.py:
```diff
@@ -67,47 +73,28 @@ def get_error_for_client(errors: List) -> str:
-    return gateway_response
+
+
+def get_template():
```
Moved to `TEMPLATE_PATH`

maarcingebala commented:
Since this PR is pretty large, I'm going to merge it as it is. There are some review comments by @patrys that are still unresolved but I decided to open new issues for them and do them separately. Those changes are already really good and I don't want them to become outdated.
